File: lectures/lecture10/code/lecture10.py
"""
MANG3095 Advanced Financial Econometrics - Week 10
Replicating a banking panel: Gopal and Schnabl (2022, RFS 35(11), 4859-4901),
"The Rise of Finance Companies and FinTech Lenders in Small Business Lending".

This unit ships the REAL county-level aggregates from the paper's replication
package (posted as supplementary material with the article at
https://doi.org/10.1093/rfs/hhac034):

  ../../../data/csv/gopal_county.csv       one row per county: 2006 bank share,
                                           2007-2016 changes in nonbank market
                                           share and lending growth, controls
  ../../../data/csv/gopal_county_year.csv  county x year panel of UCC-filing
                                           loan counts by lender type, 2006-2016

The loan-level UCC data are proprietary and are NOT included; the county
aggregates above are the paper's own published analysis files. On this data the
script reproduces Table 4 EXACTLY: 0.212 / 0.534 / 0.016 in the preferred
specification (controls + state fixed effects, county-clustered SEs).

Requires: numpy, pandas, statsmodels (all in the module's Python distribution).
Run from inside code/:  python lecture10.py   (must exit 0)
"""

import numpy as np
import pandas as pd
import statsmodels.api as sm
import statsmodels.formula.api as smf

DATA = "../../../data/csv/"

CONTROLS = ["unemp_rate_2002_2006", "lfp_rate_2002_2006", "est_2002_2006",
            "wage_2002_2006", "log_pop", "unemp_rate2005", "lfp_rate2005",
            "wage2005"]
CTRL = " + ".join(CONTROLS)

print("=" * 72)
print("1. THE REAL PANEL: STRUCTURE, IDENTIFIERS AND THE m:1 MERGE")
print("=" * 72)

cy = pd.read_csv(DATA + "gopal_county_year.csv")        # county x year counts
cs = pd.read_csv(DATA + "gopal_county.csv")             # county cross-section
cs["state"] = cs["state"].astype(str).str.zfill(2)

print(f"county-year panel: N = {len(cy)}  counties = {cy['county'].nunique()}"
      f"  years = {cy['year'].min()}-{cy['year'].max()}")
counts = cy.groupby("county")["year"].size()
nyears = cy["year"].nunique()
print(f"balanced counties: {(counts == nyears).sum()} of {counts.size}"
      f" ({100 * (counts == nyears).mean():.1f}%) -> mildly unbalanced, keep all")

# the package's m:1 merge pattern: many county-years to one county record
cy = cy.merge(cs[["county", "state", "bank_share_06"]], on="county", how="inner")
cy["nonbank_share"] = cy["nonbank_loans"] / (cy["nonbank_loans"] + cy["bank_loans"])
print(f"after m:1 merge on county id: N = {len(cy)}"
      f"  (every county-year now carries its 2006 bank share)")
print(cy[["county", "state", "year", "nonbank_loans", "bank_loans",
          "nonbank_share", "bank_share_06"]].head(4).to_string(index=False))

print()
print("=" * 72)
print("2. TABLE 4, FOR REAL (long difference 2007-2016)")
print("=" * 72)

an = cs.dropna(subset=CONTROLS + ["d_nb_share_07_16", "nb_growth_07_16",
                                  "total_growth_07_16", "bank_share_06"]).copy()
print(f"analysis sample: {len(an)} counties, {an['state'].nunique()} states")
print(f"bank share 2006: mean {an['bank_share_06'].mean():.3f}"
      f"  sd {an['bank_share_06'].std():.3f}   (paper Table 3: 0.476, 0.130)")

def cluster_fit(formula, data, cluster="county"):
    cols = [c for c in data.columns if c in formula] + [cluster]
    d = data.dropna(subset=list(set(cols))).copy()
    return smf.ols(formula, data=d).fit(
        cov_type="cluster", cov_kwds={"groups": d[cluster]}), d

OUTCOMES = [("d_nb_share_07_16", "Nonbank market share"),
            ("nb_growth_07_16",  "Nonbank lending     "),
            ("total_growth_07_16", "Total lending       ")]

print("\ncoefficient on BankShare_06 (county-clustered SE):")
print(f"{'outcome':24s} {'no controls':>16s} {'controls':>16s} {'ctrls+state FE':>16s}")
pref = {}
for yvar, label in OUTCOMES:
    row = []
    for rhs in (f"{yvar} ~ bank_share_06",
                f"{yvar} ~ bank_share_06 + {CTRL}",
                f"{yvar} ~ bank_share_06 + {CTRL} + C(state)"):
        m, _ = cluster_fit(rhs, an)
        row.append((m.params["bank_share_06"], m.bse["bank_share_06"]))
    pref[yvar] = row[2]
    print(f"{label:24s}" + "".join(f"   {b: .3f} ({s:.3f})" for b, s in row))
print("paper, preferred spec:      0.212 (0.024)   0.534 (0.080)   0.016 (0.060)")

g, g_se = pref["d_nb_share_07_16"]
p10, p90 = an["bank_share_06"].quantile([0.10, 0.90])
print(f"\n10th->90th pct of bank share ({p10:.3f} -> {p90:.3f}) moves the nonbank")
print(f"market share by {g * (p90 - p10) * 100:+.1f} pp; total lending is a precise null.")

print()
print("=" * 72)
print("3. THE CRISIS EVENT STUDY ON THE REAL PANEL (county + year FE)")
print("=" * 72)

# nonbank share, county FE removed by within-demeaning, interacted year effects:
# gamma_t on BankShare_06 x 1[year=t], base year 2007 (the paper's Figure 4 logic)
# 2006 is excluded: the exposure measure is BUILT from the 2006 filing counts,
# so its 2006 cross-section slope is -1 mechanically (a circularity, not a trend)
evp = cy[cy["year"] >= 2007].copy()
evp["nb_dm"] = evp["nonbank_share"] - evp.groupby("county")["nonbank_share"].transform("mean")
evd = evp.dropna(subset=["nb_dm", "bank_share_06"]).copy()
ev = smf.ols("nb_dm ~ C(year):bank_share_06 + C(year)", data=evd).fit(
    cov_type="cluster", cov_kwds={"groups": evd["county"]})
basecoef = ev.params["C(year)[2007]:bank_share_06"]
print("gamma_t on BankShare_06 x year (relative to 2007), nonbank share outcome:")
for yy in range(2007, 2017):
    c = ev.params[f"C(year)[{yy}]:bank_share_06"] - basecoef
    se = ev.bse[f"C(year)[{yy}]:bank_share_06"]
    bar = "#" * int(round(abs(c) * 200))
    print(f"  {yy}: {c:+.3f} (se {se:.3f}) {bar}")
print("near zero through 2009, then rising steadily: high-bank-share counties")
print("see nonbanks take over after the bank shock, and the gap never closes.")

print()
print("=" * 72)
print("4. CLUSTERED STANDARD ERRORS ON THE REAL REGRESSION")
print("=" * 72)

m_pref, d_pref = cluster_fit(f"d_nb_share_07_16 ~ bank_share_06 + {CTRL} + C(state)", an)
plain = smf.ols(f"d_nb_share_07_16 ~ bank_share_06 + {CTRL} + C(state)",
                data=d_pref).fit()
se_iid = plain.bse["bank_share_06"]
se_cty = plain.get_robustcov_results(
    cov_type="cluster", groups=d_pref["county"]).bse[
    list(plain.params.index).index("bank_share_06")]
se_st = plain.get_robustcov_results(
    cov_type="cluster", groups=d_pref["state"]).bse[
    list(plain.params.index).index("bank_share_06")]
print(f"gamma = {plain.params['bank_share_06']:.3f}; standard error by treatment of the errors:")
print(f"  iid (naive)     {se_iid:.4f}")
print(f"  county cluster  {se_cty:.4f}   (the paper's choice)")
print(f"  state cluster   {se_st:.4f}   ({d_pref['state'].nunique()} clusters: coarser, noisier)")
print(f"naive/county ratio = {se_cty / se_iid:.2f}: even in a cross-section the")
print("errors are not iid, and the panel event study above would be far worse.")

# Monte Carlo demonstration of the Moulton problem (placebo county-level
# regressor, errors correlated within county; this is a statistics
# demonstration by construction, not paper data)
REPS = 1000
rej_naive = rej_clust = 0
mc = np.random.default_rng(2024)
for _ in range(REPS):
    Gc, Tc, icc = 40, 8, 0.5
    x_c = mc.normal(0, 1, Gc)
    u = (np.sqrt(icc) * np.repeat(mc.normal(0, 1, Gc), Tc)
         + np.sqrt(1 - icc) * mc.normal(0, 1, Gc * Tc))
    yv = u
    xv = np.repeat(x_c, Tc)
    ids = np.repeat(np.arange(Gc), Tc)
    n = len(yv)
    xd = xv - xv.mean()
    b = (xd @ (yv - yv.mean())) / (xd @ xd)
    e = (yv - yv.mean()) - b * xd
    se_n = np.sqrt((e @ e) / (n - 2) / (xd @ xd))
    scores = np.array([xd[ids == gg] @ e[ids == gg] for gg in range(Gc)])
    se_c = np.sqrt((scores @ scores) / (xd @ xd) ** 2 * Gc / (Gc - 1))
    rej_naive += abs(b / se_n) > 1.96
    rej_clust += abs(b / se_c) > 2.02          # t(39) 5% critical value
print(f"placebo Monte Carlo ({REPS} draws, county regressor, true effect 0):")
print(f"  naive OLS rejects H0      : {100 * rej_naive / REPS:.1f}% at a nominal 5%")
print(f"  cluster-robust rejects H0 : {100 * rej_clust / REPS:.1f}%")

print()
print("=" * 72)
print("5. ROBUSTNESS ON THE REAL DATA")
print("=" * 72)

# (a) subperiods: the effect builds after 2010
for yvar, lab in [("d_nb_share_07_10", "2007-2010"), ("d_nb_share_10_16", "2010-2016")]:
    m, _ = cluster_fit(f"{yvar} ~ bank_share_06 + {CTRL} + C(state)", an)
    print(f"  Delta nonbank share {lab}: gamma = {m.params['bank_share_06']:.3f}"
          f" (se {m.bse['bank_share_06']:.3f})")
print("  the gap keeps widening long after 2010: a permanent reallocation,")
print("  not a temporary crisis dip (the paper's Figure 4 message).")

# (b) leave-one-state-out: is gamma driven by any single state?
gammas = []
for s in sorted(d_pref["state"].unique()):
    sub = d_pref[d_pref["state"] != s]
    mm = smf.ols(f"d_nb_share_07_16 ~ bank_share_06 + {CTRL} + C(state)",
                 data=sub).fit()
    gammas.append(mm.params["bank_share_06"])
print(f"  leave-one-state-out gamma: min {min(gammas):.3f}, max {max(gammas):.3f}"
      f"  (full sample {g:.3f})")

# (c) specification ladder recap
print(f"  no controls {0.215:.3f} -> controls {0.197:.3f} -> +state FE {g:.3f}:")
print("  stable across specifications, the sign of a design, not a regression.")

print()
print("Done. All numbers above come from the paper's own replication data;")
print("Table 4's 0.212 / 0.534 / 0.016 reproduce exactly.")

File: lectures/lecture10/code/lecture10.R
# ===========================================================================
# MANG3095 Advanced Financial Econometrics - Week 10
# Replicating a banking panel: Gopal and Schnabl (2022, RFS 35(11), 4859-4901)
# "The Rise of Finance Companies and FinTech Lenders in Small Business Lending"
# ===========================================================================
# This script runs on the REAL county-level aggregates from the paper's
# replication package (supplementary material, doi.org/10.1093/rfs/hhac034):
#   ../../../data/csv/gopal_county.csv       county cross-section: 2006 bank
#                                            share, 2007-2016 changes, controls
#   ../../../data/csv/gopal_county_year.csv  county x year UCC loan counts
# On this data the preferred specification reproduces Table 4 EXACTLY:
#   0.212 (0.024) / 0.534 (0.080) / 0.016 (0.060)
# Packages: install.packages(c("fixest", "dplyr", "tidyr"))
#   fixest::feols() is the R twin of the package's reghdfe: the formula
#   y ~ x | fe1 + fe2 absorbs FE and cluster = ~id clusters the SEs.
# ===========================================================================

library(dplyr)
library(fixest)

DATA <- "../../../data/csv/"

controls <- c("unemp_rate_2002_2006", "lfp_rate_2002_2006", "est_2002_2006",
              "wage_2002_2006", "log_pop", "unemp_rate2005", "lfp_rate2005",
              "wage2005")
CTRL <- paste(controls, collapse = " + ")

# ---------------------------------------------------------------------------
# 1. The real panel: structure, identifiers and the m:1 merge
# ---------------------------------------------------------------------------

cy <- read.csv(paste0(DATA, "gopal_county_year.csv"))
cs <- read.csv(paste0(DATA, "gopal_county.csv"))

cat("county-year panel:", nrow(cy), "rows;", n_distinct(cy$county),
    "counties;", min(cy$year), "-", max(cy$year), "\n")
tab <- table(table(cy$county))
cat("balanced counties (all", n_distinct(cy$year), "years):",
    sum(table(cy$county) == n_distinct(cy$year)), "of",
    n_distinct(cy$county), "-> mildly unbalanced, keep all\n")

# the package's m:1 merge pattern: many county-years to one county record
cy <- cy |>
  inner_join(cs |> select(county, state, bank_share_06), by = "county") |>
  mutate(nonbank_share = nonbank_loans / (nonbank_loans + bank_loans))
head(cy, 4)

# ---------------------------------------------------------------------------
# 2. Table 4, for real (long difference 2007-2016)
# ---------------------------------------------------------------------------
# The package's line:  reghdfe share_07_16 pre_crisis_bank_share `controls'
#                          if bank==0, abs(state) cluster(county)

an <- cs |> filter(complete.cases(across(all_of(c(controls, "d_nb_share_07_16",
        "nb_growth_07_16", "total_growth_07_16", "bank_share_06")))))
cat("analysis sample:", nrow(an), "counties;",
    "bank share mean", round(mean(an$bank_share_06), 3),
    "sd", round(sd(an$bank_share_06), 3), "(paper Table 3: 0.476, 0.130)\n")

f <- function(y, fe = TRUE, ctrl = TRUE) {
  rhs <- paste0(y, " ~ bank_share_06", if (ctrl) paste0(" + ", CTRL) else "",
                if (fe) " | state" else "")
  feols(as.formula(rhs), an, cluster = ~county)
}
m1 <- f("d_nb_share_07_16");  m2 <- f("nb_growth_07_16");  m3 <- f("total_growth_07_16")
etable(f("d_nb_share_07_16", fe = FALSE, ctrl = FALSE),
       f("d_nb_share_07_16", fe = FALSE), m1, m2, m3,
       keep = "bank_share_06",
       headers = c("share raw", "share ctrl", "share FE", "nb growth", "total"))
# expected FE row: 0.212 (0.024)   0.534 (0.080)   0.016 (0.060)

q <- quantile(an$bank_share_06, c(0.10, 0.90))
cat("10th->90th pct of bank share moves the nonbank share by",
    round(coef(m1)["bank_share_06"] * diff(q) * 100, 1),
    "pp; total lending is a precise null\n")

# ---------------------------------------------------------------------------
# 3. The crisis event study on the real panel (county + year FE)
# ---------------------------------------------------------------------------
# 2006 is excluded: the exposure measure is built FROM the 2006 counts,
# so its 2006 slope is -1 mechanically. Base year 2007.

ev <- cy |> filter(year >= 2007, !is.na(nonbank_share))
m_ev <- feols(nonbank_share ~ i(year, bank_share_06, ref = 2007) | county + year,
              ev, cluster = ~county)
print(coef(m_ev))
# expected path: +0.04 (2008), +0.02 (2009), +0.06 (2010) ... +0.22 (2016)
# (fixest::iplot(m_ev) draws the event-study path)

# ---------------------------------------------------------------------------
# 4. Clustered standard errors on the real regression
# ---------------------------------------------------------------------------

m_iid <- summary(m1, vcov = "iid")
m_cty <- summary(m1, cluster = ~county)
m_st  <- summary(m1, cluster = ~state)
cat("gamma se: iid", round(se(m_iid)["bank_share_06"], 4),
    "| county", round(se(m_cty)["bank_share_06"], 4),
    "| state", round(se(m_st)["bank_share_06"], 4),
    "(50 clusters: coarser, noisier; Cameron-Miller 2015)\n")
# expected: 0.019 / 0.024 / 0.027 - even a cross-section is not iid

# ---------------------------------------------------------------------------
# 5. Robustness on the real data
# ---------------------------------------------------------------------------

# (a) subperiods: the effect builds after 2010
etable(f("d_nb_share_07_10"), f("d_nb_share_10_16"), keep = "bank_share_06",
       headers = c("2007-2010", "2010-2016"))
# expected: 0.073 (0.023) then 0.138 (0.024): a permanent reallocation

# (b) leave-one-state-out: no single state drives gamma
states <- unique(an$state)
gammas <- sapply(states, function(s)
  coef(feols(as.formula(paste0("d_nb_share_07_16 ~ bank_share_06 + ", CTRL,
                               " | state")), an[an$state != s, ]))["bank_share_06"])
cat("leave-one-state-out gamma: min", round(min(gammas), 3),
    "max", round(max(gammas), 3), " (full sample 0.212)\n")

# ===========================================================================
# What to interpret: the 2006 bank share predicts where nonbanks took over
# after 2008 (gamma = 0.212, Table 4 exactly); the event-study path lines up
# with the shock and never reverts; total lending is a precise null - lender
# substitution, not a credit crunch. County clustering matters even in the
# cross-section (se 0.019 -> 0.024).
# ===========================================================================
